virttest/libvirt_xml/libvirt_cpu.py

```diff
@@ -0,0 +1,110 @@
+    :return the updated vmxml
+    """
+    for attr, value in list(vm_attrs.items()):
+        logging.debug('Set %s = %s', attr, value)
+        setattr(vmxml, attr, int(value) if value.isdigit() else value)
+    vmxml.xmltreefile.write()
+    vmxml.sync()
+    return vmxml
+
+
+def add_cpu_settings(vmxml, params):
+    """
+    Add cpu element/value pairs to the cpu xml section and keeping the existing
+    contents if they are already in the xml. If the numa nodes is added, the
+    number of vcpus in nodes should match the number in vcpu element. So, the
+    setvm_vcpu pair need to be in params. The same for memory size(max_mem). The
+    elements can be updated are list as below:
+        setvm_max_mem_rt_slots = "16"
+        setvm_max_mem_rt = 8192
+        setvm_max_mem_rt_unit = "MiB"
+        setvm_max_mem = 1536
+        setvm_max_mem_unit = "MiB"
+        setvm_current_mem = 1536
+        setvm_current_mem_unit = "MiB"
+        setvm_vcpu = 4
+
+
+    :param vmxml: VMXML instance of the domain
+    :param params: dict of the cpu related parameter pairs
+    :return the updated vmxml
+    """
+    feature_list = None
+    if vmxml.xmltreefile.find('cpu'):
+        cpu_xml = vmxml.cpu
+        feature_list = cpu_xml.get_feature_list()
+    else:
+        cpu_xml = vm_xml.VMCPUXML()
+
+    if cpu_xml.xmltreefile.find('mode'):
+        cpu_mode = cpu_xml.mode
+    else:
+        cpu_mode = params.get("cpuxml_cpu_mode", "host-model")
+
+    if cpu_xml.xmltreefile.find("model"):
+        cpu_model = cpu_xml.model
+    else:
+        cpu_model = params.get("cpuxml_model", "")
+
+    if cpu_xml.xmltreefile.find("fallback"):
+        cpu_model = cpu_xml.fallback
+    else:
+        cpu_fallback = params.get("cpuxml_fallback", "forbid")
+
+    cpu_xml_new = vm_xml.VMCPUXML()
+    cpu_xml_new.xml = "<cpu><numa/></cpu>"
+    cpu_xml_new.mode = cpu_mode
+    cpu_xml_new.model = cpu_model
+    cpu_xml_new.fallback = cpu_fallback
+
+    feature_name_add = params.get("cpu_feature", None)
+    feature_policy_add = params.get("cpu_feature_policy", None)
+    if feature_list:
+        for i in range(0, len(feature_list)):
+            feature_name = cpu_xml.get_feature(i).get('name')
+            if feature_name == feature_name_add:
+                feature_name_add = ""
+            feature_policy = cpu_xml.get_feature(i).get('policy')
+            cpu_xml_new.add_feature(feature_name, feature_policy)
+    if feature_name_add and feature_policy_add:
+        cpu_xml_new.add_feature(feature_name_add, feature_policy_add)
+
+    if cpu_xml.xmltreefile.find('numa'):
+        cpu_xml_new = cpu_xml.get_numa_cell()
+    else:
+        cells = eval(params.get("cpuxml_numa_cell", "[]"))
+        numacell_xmls = []
+        for i in range(len(cells)):
+            numacell_xml = vm_xml.NumaCellXML()
+            numacell_xml.update(cells[i])
+            numacell_xmls.append(numacell_xml)
+        cpu_xml.numa_cell = numacell_xmls
```

Comment:
With PR https://github.com/avocado-framework/avocado-vt/pull/2749 merged, maybe we are able to simplify the usage for numa cell setting. Could you help refer https://github.com/autotest/tp-libvirt/pull/3057 to see if we can refine this PR?

Reply:
Updated